tensorflow/core/platform/windows/windows_file_system.cc

```diff
@@ -136,131 +226,262 @@ Status WindowsFileSystem::NewRandomAccessFile(
     const string& fname, std::unique_ptr<RandomAccessFile>* result) {
   string translated_fname = TranslateName(fname);
   result->reset();
-  Status s;
-  FILE* f = fopen(translated_fname.c_str(), "r");
-  if (f == NULL) {
-    s = IOError(fname, errno);
-  } else {
-    result->reset(new WindowsRandomAccessFile(translated_fname, f));
+
+  // Open the file for read-only random access
+  // Random access is to disable read-ahead as the system reads too much data
+  // Open in async mode which makes Windows allow more parallelism even
+  // if we need to do sync I/O on top of it.
+  DWORD file_flags = FILE_ATTRIBUTE_READONLY | FILE_FLAG_RANDOM_ACCESS |
+      FILE_FLAG_OVERLAPPED;
+  // Shared access is necessary for tests to pass
+  // almost all tests would work with a possible exception of fault_injection.
+  DWORD share_mode = FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE;
+
+  HANDLE hfile = ::CreateFileA(translated_fname.c_str(), GENERIC_READ,
+                               share_mode, NULL, OPEN_EXISTING, file_flags,
+                               NULL);
+
+  if (INVALID_HANDLE_VALUE == hfile) {
+    string context = "NewRandomAccessFile failed to Create/Open: " + fname;
+    return IOErrorFromWindowsError(context, ::GetLastError());
   }
-  return s;
+
+  UniqueCloseHandlePtr file_guard(hfile, CloseHandleFunc);
+  result->reset(new WindowsRandomAccessFile(translated_fname, hfile));
+  file_guard.release();
+
+  return Status::OK();
 }
 
 Status WindowsFileSystem::NewWritableFile(
     const string& fname, std::unique_ptr<WritableFile>* result) {
   string translated_fname = TranslateName(fname);
-  Status s;
-  FILE* f = fopen(translated_fname.c_str(), "w");
-  if (f == NULL) {
-    result->reset();
-    s = IOError(fname, errno);
-  } else {
-    result->reset(new WindowsWritableFile(translated_fname, f));
+  result->reset();
+
+  DWORD share_mode = FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE;
+  HANDLE hfile = ::CreateFileA(translated_fname.c_str(), GENERIC_WRITE,
+                               share_mode, NULL, CREATE_ALWAYS,
+                               FILE_ATTRIBUTE_NORMAL, NULL);
+
+  if (INVALID_HANDLE_VALUE == hfile) {
+    string context = "Failed to create a NewWriteableFile: " + fname;
+    return IOErrorFromWindowsError(context, ::GetLastError());
   }
-  return s;
+
+  UniqueCloseHandlePtr file_guard(hfile, CloseHandleFunc);
```

Comment:
Does the `UniqueCloseHandlePtr` serve a purpose here? (I would expect it to be used when there is the possibility of returning early, due to an error, but it looks like L272 can't fail, so we are just creating and destroying the guard for no purpose. Perhaps I'm missing something.)